File: kiosk/src/lib/features.ts
/** The kiosk's feature registry: what the Home launcher tiles link to,
 *  what routes exist, and what the top bar's section label reads.
 *  Permission gating of tiles is deferred (see spec). Tile availability
 *  by kiosk setup state is handled here via `alwaysAvailable` /
 *  `featureAvailable` — see `setupState.ts`. */

import { isSetupComplete, type KioskSetupState } from './setupState';

export interface KioskFeature {
  id: 'setup' | 'scan' | 'labels' | 'timeclock' | 'settings';
  path: string;
  title: string;
  blurb: string;
  /** True for features that still fall back to the generic FeaturePage
   *  placeholder; absent for features with a real screen (Settings,
   *  Kiosk Setup, and Scanning). */
  placeholder?: boolean;
  /** True for the two tiles that stay usable no matter the kiosk setup
   *  state (Kiosk Setup and Settings) — see `featureAvailable`. */
  alwaysAvailable?: boolean;
}

export const FEATURES: KioskFeature[] = [
  { id: 'setup', path: '/setup', title: 'Kiosk Setup', blurb: 'Set up this kiosk for a move.', alwaysAvailable: true },
  { id: 'scan', path: '/scan', title: 'Scanning', blurb: 'Scan assets, containers, and badges.' },
  { id: 'labels', path: '/labels', title: 'Label Printing', blurb: 'Print asset and container labels.', placeholder: true },
  { id: 'timeclock', path: '/timeclock', title: 'Timeclock', blurb: 'Clock in and out of a move.', placeholder: true },
  { id: 'settings', path: '/settings', title: 'Settings', blurb: 'Appearance, sound, devices, and more.', alwaysAvailable: true },
];

/** Whether a tile/route may be used given the kiosk's setup state — always
 *  true for `alwaysAvailable` features (Kiosk Setup, Settings) or when
 *  developer mode is on (Jimmy: "developer mode overrides this to always
 *  allow all options"), otherwise only once setup is complete. */
export function featureAvailable(
  feature: KioskFeature,
  setupState: KioskSetupState,
  devMode = false,
): boolean {
  return devMode || feature.alwaysAvailable || isSetupComplete(setupState);
}

File: Android_Kiosk_App/app/src/main/java/com/serversherpa/kiosk/core/features/Features.kt
package com.serversherpa.kiosk.core.features

import com.serversherpa.kiosk.core.setup.SetupState

enum class FeatureId { SETUP, SCAN, ENROLL, CONTAINERS, TRUCKS, LABELS, TIMECLOCK, SETTINGS }

/** One launcher tile / route. `placeholder` features open the generic
 *  "not available yet" page; `alwaysAvailable` ones ignore setup state. */
data class KioskFeature(
    val id: FeatureId,
    val route: String,
    val title: String,
    val blurb: String,
    val placeholder: Boolean = false,
    val alwaysAvailable: Boolean = false,
)

/** Same order and copy as kiosk/src/lib/features.ts. */
val FEATURES: List<KioskFeature> = listOf(
    KioskFeature(FeatureId.SETUP, "setup", "Kiosk Setup", "Set up this kiosk for a move.", alwaysAvailable = true),
    KioskFeature(FeatureId.SCAN, "scan", "Scanning", "Scan assets, containers, and badges."),
    KioskFeature(FeatureId.ENROLL, "enroll", "RFID Enroll", "Scan an asset, then scan its RFID tag."),
    KioskFeature(FeatureId.CONTAINERS, "containers", "Containers", "Pack and unpack containers by scanning.", placeholder = true),
    KioskFeature(FeatureId.TRUCKS, "trucks", "Trucks", "Load and unload trucks by scanning.", placeholder = true),
    KioskFeature(FeatureId.LABELS, "labels", "Label Printing", "Print asset and container labels.", placeholder = true),
    KioskFeature(FeatureId.TIMECLOCK, "timeclock", "Timeclock", "Clock in and out of a move."),
    KioskFeature(FeatureId.SETTINGS, "settings", "Settings", "Appearance, sound, devices, and more.", alwaysAvailable = true),
)

fun feature(id: FeatureId): KioskFeature = FEATURES.first { it.id == id }

/** The feature whose route the nav destination is (route may carry `?tab=`). */
fun featureForRoute(route: String?): KioskFeature? {
    val base = route?.substringBefore('?')?.substringBefore('/') ?: return null
    return FEATURES.firstOrNull { it.route == base }
}

/** Always for alwaysAvailable features or in dev mode; otherwise only once setup is complete. */
fun featureAvailable(feature: KioskFeature, setupState: SetupState, devMode: Boolean = false): Boolean =
    devMode || feature.alwaysAvailable || setupState.isComplete
